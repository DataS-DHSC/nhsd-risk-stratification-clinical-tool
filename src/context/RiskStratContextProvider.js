/* eslint-disable max-len */
/* eslint-disable react/prop-types */
import React, { useState } from 'react';

const RiskStratContextProvider = ({
  children,
  RiskStratContext,
  initialState,
}) => {
  let initialContextState = null;
  const isRunningClientSide = typeof window !== 'undefined';
  if (isRunningClientSide) {
    // we only want to use the context state if this code is running client side
    // when javascript is disabled, this code will run server side
    // in that case, we want to return null for the bmaData to force pages to use POST prevData state rather than the context state
    initialContextState = {
      ...initialState,
    };
  }

  const [contextData, setContextData] = useState(initialContextState);

  const setField = (field, value) => {
    setContextData({ ...contextData, [field]: value });
  };

  const resetContext = () => {
    setContextData({ ...initialState });
  };

  return (
    <RiskStratContext.Provider value={[contextData, setField, resetContext]}>
      {children}
    </RiskStratContext.Provider>
  );
};

export default RiskStratContextProvider;
