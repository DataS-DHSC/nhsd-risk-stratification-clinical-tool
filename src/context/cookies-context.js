import { createContext, useContext, useState } from 'react';
import PropTypes from 'prop-types';

function parseCookie(str) {
  switch (str) {
    case 'true':
      return true;
    case 'false':
      return false;
    case undefined:
      return undefined;
    default:
      // eslint-disable-next-line no-console
      console.error(`Unable to parse opt-in value ${str}`);
      return undefined;
  }
}

export const CookiesContext = createContext({
  optIn: undefined,
  setOptIn: () => {
    throw Error(
      'setOptIn function not defined, make sure to use context provider'
    );
  },
  hideSuccessBanner: () => {
    throw Error(
      'hideSuccessBanner function not defined, make sure to use context provider'
    );
  },
  showSuccessBanner: () => {
    throw Error(
      'showSuccessBanner function not defined, make sure to use context provider'
    );
  },
  isSuccessBannerVisible: undefined,
  isSelectionBannerVisible: () => {
    throw Error(
      'isSelectionBannerVisible function not defined, make sure to use context provider'
    );
  },
});

export const useCookiesContext = () => useContext(CookiesContext);

export default function CookiesContextProvider({ children, getCookie }) {
  const [optIn, setOptIn] = useState(parseCookie(getCookie('opt-in')));

  const [isSuccessBannerVisible, setIsSuccessBannerVisible] = useState(false);
  const hideSuccessBanner = () => setIsSuccessBannerVisible(false);
  const showSuccessBanner = () => setIsSuccessBannerVisible(true);

  const isSelectionBannerVisible = () => optIn === undefined;

  return (
    <CookiesContext.Provider
      value={{
        optIn,
        setOptIn,
        hideSuccessBanner,
        showSuccessBanner,
        isSuccessBannerVisible,
        isSelectionBannerVisible,
      }}
    >
      {children}
    </CookiesContext.Provider>
  );
}

CookiesContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
  getCookie: PropTypes.func.isRequired,
};
