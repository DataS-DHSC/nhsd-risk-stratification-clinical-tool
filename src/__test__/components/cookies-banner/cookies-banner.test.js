import { render } from '@testing-library/react';
import { RouterContext } from 'next/dist/next-server/lib/router-context';
import Router from 'next/router';
import CookiesBanner from '../../../components/cookies-banner/cookies-banner';
import { CookiesContext } from '../../../context/cookies-context';

const setup = (contextProviderValue) =>
  render(
    <RouterContext.Provider value={Router}>
      <CookiesContext.Provider value={contextProviderValue}>
        <CookiesBanner />
      </CookiesContext.Provider>
    </RouterContext.Provider>
  );

describe('Cookies Banner', () => {
  it('should display SelectionBanner', () => {
    const contextProviderValue = {
      optIn: undefined,
      isSuccessBannerVisible: false,
      isSelectionBannerVisible: () => true,
    };
    const container = setup(contextProviderValue);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should display SuccessBanner', () => {
    const contextProviderValue = {
      optIn: true,
      isSuccessBannerVisible: true,
      isSelectionBannerVisible: () => false,
    };
    const container = setup(contextProviderValue);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should display nothing', () => {
    const contextProviderValue = {
      optIn: true,
      isSuccessBannerVisible: false,
      isSelectionBannerVisible: () => false,
    };
    const container = setup(contextProviderValue);
    expect(container.asFragment()).toMatchSnapshot();
  });
});
