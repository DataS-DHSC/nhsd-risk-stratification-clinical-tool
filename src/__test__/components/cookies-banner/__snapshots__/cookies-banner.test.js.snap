// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cookies Banner should display SelectionBanner 1`] = `
<DocumentFragment>
  <nav
    class="nhsuk-cookie-banner"
    id="nhsuk-cookie-banner"
  >
    <div
      class="nhsuk-width-container"
    >
      <h3>
        Cookies on the NHS website
      </h3>
      <p>
        We've put some small files called cookies on your device to make our site work.
      </p>
      <p>
        We'd also like to use analytics cookies. These send information about how our site is used to services called Adobe Analytics, Hotjar and Google Analytics. We use this information to improve our site.
      </p>
      <p>
        Let us know if this is OK. We'll use a cookie to save your choice. You can 
        <a
          href="/page/cookies/policy"
          id="nhsuk-cookie-banner__link"
        >
          read more about our cookies
        </a>
         before you choose.
      </p>
      <ul>
        <li>
          <button
            class="nhsuk-button"
            id="nhsuk-cookie-banner-opt-in"
            type="submit"
          >
            I'm OK with analytics cookies
          </button>
        </li>
        <li>
          <button
            class="nhsuk-button"
            id="nhsuk-cookie-banner-opt-out"
            type="submit"
          >
            Do not use analytics cookies
          </button>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`Cookies Banner should display SuccessBanner 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-success-banner"
    id="nhsuk-success-banner"
  >
    <div
      class="nhsuk-width-container"
    >
      <p
        class="nhsuk-success-cookie-banner__message"
        id="nhsuk-success-banner__message"
      >
        You can change your cookie settings at any time using our 
        <a
          href="/page/cookies/policy"
        >
          cookies page
        </a>
        .
      </p>
      <p
        class="nhsuk-success-cookie-banner__message"
        id="nhsuk-success-banner-hide__message"
      >
        <a
          href=""
        >
          Hide this message
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cookies Banner should display nothing 1`] = `<DocumentFragment />`;
